packages/@ember/-internals/glimmer/lib/component-managers/custom.ts

```diff
@@ -150,16 +154,82 @@ export default class CustomComponentManager<ComponentInstance>
     const { delegate } = definition;
     const capturedArgs = args.capture();
 
-    const component = delegate.createComponent(
-      definition.ComponentClass.class,
-      capturedArgs.value()
-    );
+    let value;
+    let namedArgsProxy = {};
+
+    if (EMBER_CUSTOM_COMPONENT_ARG_PROXY) {
+      if (HAS_NATIVE_PROXY) {
+        let handler: ProxyHandler<{}> = {
+          get(_target, prop) {
+            assert('args can only be strings', typeof prop === 'string');
+
+            let tracker = getCurrentTracker();
+            let ref = capturedArgs.named.get(prop as string);
+
+            if (tracker) {
+              tracker.add(ref.tag);
+            }
+
+            return ref.value();
+          },
+        };
+
+        if (DEBUG) {
+          handler.set = function(target, prop) {
+            assert(
+              `You attempted to set ${target}#${String(
+                prop
+              )} on a components arguments. Component arguments are immutable and cannot be updated directly, they always represent the values that are passed to your component. If you want to set default values, you should use a getter instead`
+            );
+
+            return false;
+          };
+        }
+
+        namedArgsProxy = new Proxy(namedArgsProxy, handler);
+      } else {
+        capturedArgs.named.names.forEach(name => {
+          Object.defineProperty(namedArgsProxy, name, {
+            get() {
+              let ref = capturedArgs.named.get(name);
+              let tracker = getCurrentTracker();
+
+              if (tracker) {
+                tracker.add(ref.tag);
+              }
+
+              return ref.value();
+            },
+          });
+        });
+      }
+
+      value = {
+        named: namedArgsProxy,
+        positional: capturedArgs.positional.value(),
+      };
+    } else {
+      value = capturedArgs.value();
+    }
 
-    return new CustomComponentState(delegate, component, capturedArgs);
+    const component = delegate.createComponent(definition.ComponentClass.class, value);
+
+    return new CustomComponentState(delegate, component, capturedArgs, namedArgsProxy);
   }
 
-  update({ delegate, component, args }: CustomComponentState<ComponentInstance>) {
-    delegate.updateComponent(component, args.value());
+  update({ delegate, component, args, namedArgsProxy }: CustomComponentState<ComponentInstance>) {
+    let value;
+
+    if (EMBER_CUSTOM_COMPONENT_ARG_PROXY) {
+      value = {
```

Comment:
with EMBER_CUSTOM_COMPONENT_ARG_PROXY you shouldn't need to recreate this or emit the update hook, it could set the updateHook capability to false

Reply:
actually can just make it so the glimmer component opts out of this hook in its manager if the EMBER_CUSTOM_COMPONENT_ARG_PROXY is set

Reply:
Yeah, it seems reasonable to add the capability to now. I wasn't sure whether or not we had a process for adding capabilities.